### PR TITLE
Integration benchmark tests

### PR DESCRIPTION
**Pull request by WojciechNagorski**
## 2023.0.0
SshClient
| Method  | Mean     | Error    | StdDev   | Allocated |
|-------- |---------:|---------:|---------:|----------:|
| Connect | 93.18 ms | 1.816 ms | 2.162 ms | 165.78 KB |

## Master
SshClient 
| Method  | Mean     | Error    | StdDev   | Allocated |
|-------- |---------:|---------:|---------:|----------:|
| Connect | 92.74 ms | 1.645 ms | 1.285 ms | 158.17 KB |

## After change

SftpClient on this branch:

| Method            | Mean     | Error    | StdDev   | Allocated |
|------------------ |---------:|---------:|---------:|----------:|
| Connect           | 95.89 ms | 1.907 ms | 2.735 ms | 190.51 KB |
| ConnectAsync      | 95.91 ms | 1.877 ms | 2.867 ms |  192.3 KB |
| UploadAndDownload | 16.48 ms | 0.236 ms | 0.221 ms |  54.45 KB |

SshClient on this branch:
| Method                    | Mean      | Error     | StdDev    | Allocated |
|-------------------------- |----------:|----------:|----------:|----------:|
| Connect                   | 35.644 ms | 0.4266 ms | 0.3782 ms | 140.41 KB |
| ConnectAsync              | 35.572 ms | 0.2993 ms | 0.2653 ms | 142.17 KB |
| ConnectAndRunCommand      | 92.598 ms | 1.7199 ms | 1.9807 ms | 156.12 KB |
| ConnectAsyncAndRunCommand | 93.462 ms | 1.8312 ms | 2.7408 ms | 157.86 KB |
| RunCommand                |  6.242 ms | 0.0646 ms | 0.0604 ms |  77.46 KB |

ScpClient on this branch:
| Method                   | Mean      | Error    | StdDev   | Allocated |
|------------------------- |----------:|---------:|---------:|----------:|
| Connect                  |  35.89 ms | 0.389 ms | 0.345 ms | 140.51 KB |
| ConnectAsync             |  35.77 ms | 0.441 ms | 0.391 ms | 142.14 KB |
| ConnectUploadAndDownload | 109.93 ms | 2.189 ms | 2.342 ms | 184.81 KB |
| UploadAndDownload        |  22.89 ms | 0.432 ms | 0.404 ms |  42.46 KB |

### CODE REVIEW

**Comment by WojciechNagorski** [comment]
Of course the times are not deterministic, but they are good enough.  However, allocations are 100 percent deterministic.

File: src/Renci.SshNet/Common/SshDataStream.cs [hunk 1]
```diff
@@ -126,9 +126,17 @@ public void Write(string s, Encoding encoding)
             {
                 throw new ArgumentNullException(nameof(encoding));
             }
-
+#if NETSTANDARD2_1_OR_GREATER || NET6_0_OR_GREATER
+            ReadOnlySpan<char> value = s;
+            var count = encoding.GetByteCount(value);
+            Span<byte> bytes = stackalloc byte[count];
```
Comment (Rob-Hague) [inline src/Renci.SshNet/Common/SshDataStream.cs]: We should guard against allocating too much on the stack

```suggestion
            Span<byte> bytes = count <= 256
                ? stackalloc byte[count]
                : new byte[count]
```

**Review by Rob-Hague — COMMENTED** [review]
Cool!

I think it could be useful for getting profiles for different scenarios. I got it to work with `[EtwProfiler]` with a reference to BenchmarkDotNet.Diagnostics.Windows. Which is better EventPipeProfiler or EtwProfiler?

**Comment by WojciechNagorski** [comment]
@Rob-Hague Can you check if is ok?

**Review by Rob-Hague — APPROVED** [review]
Nice

**Comment by WojciechNagorski** [comment]
This issue has been fixed in the 2024.0.0 version.

**Pull request merged**